p2p/swarm.go

```diff
@@ -877,3 +870,48 @@ func (s *swarm) hasOutgoingPeer(peer p2pcrypto.PublicKey) bool {
 	s.outpeersMutex.RUnlock()
 	return ok
 }
+
+var listeningIp = inet.ParseIP("0.0.0.0")
+
+// TODO: test
+func (s *swarm) getListeners(
+	getTcpListener func(tcpAddr *inet.TCPAddr) (inet.Listener, error),
+	getUdpListener func(udpAddr *inet.UDPAddr) (*inet.UDPConn, error),
+) (inet.Listener, *inet.UDPConn, error) {
+
+	port := s.config.TCPPort
+	randomPort := port == 0
+
+	for {
+		tcpAddr := &inet.TCPAddr{IP: listeningIp, Port: port}
+		tcpListener, err := getTcpListener(tcpAddr)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to acquire requested tcp port: %v", err)
+		}
+
+		addr := tcpListener.Addr()
+		port = addr.(*inet.TCPAddr).Port
```

Comment:
just clarifying something, if the given port was 0 (random), and we acquired a TCP port but the equivalent UDP  was taken already. we'll loop over again to find a port which is the same for both. that wasn't the case before that change and we could assign two different random ports.

Reply:
Yes, it's as you described. We now require both TCP and UDP to be on the same port.

This is because of how the wrapper library is implemented. If we end up creating our own wrapper, we can remove this restriction.